### ISSUE TITLE
[RFC + BUG?] Implementation of Categorical distribution

### ISSUE THREAD

**Issue by dibgerge** [apache/incubator-mxnet#19722]
## Problem statement

I am probably wrong about those, but I thought to ask here anyway. I was using `mxnet.gluon.probability.Categorical` from master, and two things came to my attention:

1. It requires a `num_events` argument, which feel unnecessary, and possibly could be eliminated?
2. [Possible bug] If providing `logit`, current `log_prob` implementation assumes we are providing `log softmax(x)`. Usually, when providing `logit`, it is the linear output of the net. This is the behavior also in torch/tensorflow, where internally, the logit is scaled by `logsumexp`.

## Proposed solutions
1. Remove `num_events` argument
2. If `logit` provided in constructor, shift it by `logsumexp`

## References
- [Tensorflow implementation](https://github.com/tensorflow/probability/blob/v0.11.1/tensorflow_probability/python/distributions/categorical.py#L289)
- [Torch implementation](https://github.com/pytorch/pytorch/blob/master/torch/distributions/categorical.py#L55)

**Comment by sxjscience** [comment]
I think adding a `logsumexp` is reasonable. When the inputs are already from log_softmax, `logsumexp(x)` will be 0 so the behavior is the same as the current version. Basically, `logsumexp(log_softmax(x)) == 0`

**Comment by xidulu** [comment]
Hi @dibgerge 
The reason "num_events" is required is that we cannot call `logit.shape` or `prob.shape` in the symbolic mode, as you can see, TF would use a very complicated subroutine to get num_events in the symbol mode: https://github.com/tensorflow/probability/blob/v0.11.1/tensorflow_probability/python/distributions/categorical.py#L352
while torch does not have such issue : https://github.com/pytorch/pytorch/blob/master/torch/distributions/categorical.py#L57

**Comment by dibgerge** [comment]
> Hi @dibgerge
> The reason "num_events" is required is that we cannot call `logit.shape` or `prob.shape` in the symbolic mode, as you can see, TF would use a very complicated subroutine to get num_events in the symbol mode: https://github.com/tensorflow/probability/blob/v0.11.1/tensorflow_probability/python/distributions/categorical.py#L352
> while torch does not have such issue : https://github.com/pytorch/pytorch/blob/master/torch/distributions/categorical.py#L57
> 
> You are also right about the `logit` issue, that could be a **bug**. But MXNet currently does not have a logsumexp Op, a relative Op would be logaddexp, which is still not merged yet: #15857
> A possible solution could be applying npx.logsoftmax on the logit.

Thank you for the clarification. I am not very familiar with the symbolic code. I wonder if it is worth implementing this subroutine to calculate shape if at all possible, or it may add significant overhead. 
Also wondering, would there be performance differences between `npx.logsoftmax` and the `logaddexp` PR?

**Comment by xidulu** [comment]
MXNet should support symbolic shape like TF does (I guess), but I have zero knowledge of it, and it's probably broken or out of maintenance or not compatible with deep NumPy api or something  :-|

Regarding ur second question, the performance should roughly be the same, since they are both doing reduction operation. (Maybe `npx.logsoftmax` could be faster)

**Comment by szha** [comment]
You can use the shape_array function on arrays to retrieve shapes as an MXNet managed array (i.e. symbol/ndarray) and use that to retrieve the num_categories as a scalar array through indexing
